plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,290 @@
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.ui.PlatformUI;
+
+public class GcpLocalRunTab extends AbstractLaunchConfigurationTab {
+
+  private static final Logger logger = Logger.getLogger(GcpLocalRunTab.class.getName());
+
+  private static final String ATTRIBUTE_ACCOUNT_EMAIL =
+      "com.google.cloud.tools.eclipse.gcpEmulation.accountEmail"; //$NON-NLS-1$
+
+  private static final String PROJECT_ID_ENVIRONMENT_VARIABLE =
+      "GOOGLE_CLOUD_PROJECT"; //$NON-NLS-1$
+  private static final String SERVICE_KEY_ENVIRONMENT_VARIABLE =
+      "GOOGLE_APPLICATION_CREDENTIALS"; //$NON-NLS-1$
+
+  private final EnvironmentTab environmentTab;
+  private final IGoogleLoginService loginService;
+  private final ProjectRepository projectRepository;
+
+  private AccountSelector accountSelector;
+  private ProjectSelector projectSelector;
+  private Text serviceKeyInput;
+
+  private Image gcpIcon;
+
+  // We set up intermediary models between a run configuration and UI components for certain values,
+  // because, e.g., the account selector cannot load an email if it is not logged in. In such a
+  // case, although nothing is selected in the account selector, we should not clear the email saved
+  // in the run configuration.
+  private String accountEmailModel;
+  private String gcpProjectIdModel;
+  // To prevent updating above models when programmatically setting up UI components.
+  private boolean initializingUiValues;
+
+  private boolean activated;
+
+  public GcpLocalRunTab(EnvironmentTab environmentTab) {
+    this(environmentTab, PlatformUI.getWorkbench().getService(IGoogleLoginService.class),
+        new ProjectRepository(PlatformUI.getWorkbench().getService(IGoogleApiFactory.class)));
+  }
+
+  @VisibleForTesting
+  GcpLocalRunTab(EnvironmentTab environmentTab,
+      IGoogleLoginService loginService, ProjectRepository projectRepository) {
+    this.environmentTab = environmentTab;
+    this.loginService = loginService;
+    this.projectRepository = projectRepository;
+  }
+
+  @Override
+  public void createControl(Composite parent) {
+    Composite composite = new Composite(parent, SWT.NONE);
+
+    // Account row
+    new Label(composite, SWT.LEAD).setText(Messages.getString("label.account")); //$NON-NLS-1$
+    accountSelector = new AccountSelector(composite, loginService);
+    accountSelector.addSelectionListener(new Runnable() {
+      @Override
+      public void run() {
+        updateProjectSelector();
+
+        if (!initializingUiValues) {
+          boolean emptySelection = accountSelector.getSelectedEmail().isEmpty();
+          boolean savedEmailAvailable = accountSelector.isEmailAvailable(accountEmailModel);
+          if (!emptySelection || savedEmailAvailable) {
+            accountEmailModel = accountSelector.getSelectedEmail();
+            gcpProjectIdModel = ""; //$NON-NLS-1$
```

Comment:
I've fiddled with this issue for quite some time. That behavior would be desirable, but if is not possible to know ahead if the user will ever change back to the previous account to restore the preserved project, or the user really wanted to change the account and then just let the project not selected (in the latter case, we shouldn't provide a wrong project ID when actually running the server.) So, I think this is what we should do, unfortunately.

Reply:
Ah, I meant:

  1. I select account A1 and project P
  2. I then select A2 which has access to P

P should stay selected.

Reply:
The case is not so different, and it is difficult to make it right, because it is assumed that the UI and the model are always in sync. Suppose the user changed to A1.5 between A1 and A2. If we could magically anticipate that the user will eventually change to A2 later, we could not clear P. However, if we do not clear P when selecting A1.5 and the user applies the change and exits, we will be wrongly providing P at runtime.

Reply:
I'm not suggesting retaining the last-selected project when switching A1 &rarr; A1.5 &rarr; A2.  It makes sense that it's cleared.

That said, I'm not sure how often this will occur in practice.  It doesn't happen with any of my current projects.

Reply:
OK, let's leave this as is. To make your scenario work, the account selector listener has to be coupled with the project selector (because it can only determine the situation after all the projects of A2 are loaded), which doesn't sound worth the effort.

In the end, all the complications and shortcomings will be eliminated by #2571.